AdminServer/appscale/admin/__init__.py

```diff
@@ -0,0 +1,404 @@
+""" A server that handles application deployments. """
+
+import argparse
+import datetime
+import logging
+import socket
+import sys
+import time
+import uuid
+
+from appscale.common import appscale_info
+from appscale.common.constants import HTTPCodes
+from appscale.common.constants import LOG_FORMAT
+from appscale.common.ua_client import UAClient
+from appscale.common.ua_client import UAException
+from appscale.common.unpackaged import APPSCALE_PYTHON_APPSERVER
+from tornado import gen
+from tornado.options import options
+from tornado import web
+from tornado.escape import json_decode
+from tornado.escape import json_encode
+from tornado.ioloop import IOLoop
+from . import utils
+from . import constants
+from .constants import (
+  CustomHTTPError,
+  MAX_DEPLOY_TIME,
+  OperationTimeout,
+  REDEPLOY_WAIT,
+  VALID_RUNTIMES
+)
+from .operations_cache import OperationsCache
+
+sys.path.append(APPSCALE_PYTHON_APPSERVER)
+from google.appengine.api.appcontroller_client import AppControllerException
+
+
+# The state of each operation.
+operations = OperationsCache()
+
+
+@gen.coroutine
+def wait_for_port_assignment(operation_id, deadline, acc):
+  """ Waits until port is assigned for version.
+
+  Args:
+    operation_id: A string specifying an operation ID.
+    deadline: A float containing a unix timestamp.
+    acc: An AppControllerClient.
+  Raises:
+    OperationTimeout if the deadline is exceeded.
+  """
+  operation = operations[operation_id]
+  project_id = operation['project_id']
+
+  while True:
+    if time.time() > deadline:
+      message = 'Deploy operation took too long.'
+      operation['done'] = True
+      operation['error'] = {'message': message}
+      raise OperationTimeout(message)
+
+    try:
+      info_map = acc.get_app_info_map()
+    except AppControllerException as error:
+      logging.warning('Unable to fetch info map: {}'.format(error))
+      yield gen.sleep(1)
+      continue
+
+    try:
+      raise gen.Return(info_map[project_id]['nginx'])
+    except KeyError:
+      yield gen.sleep(1)
+      continue
+
+
+@gen.coroutine
+def wait_for_port_to_open(http_port, operation_id, deadline):
+  """ Waits until port is open.
+
+  Args:
+    http_port: An integer specifying the version's port number.
+    operation_id: A string specifying an operation ID.
+    deadline: A float containing a unix timestamp.
+  Raises:
+    OperationTimeout if the deadline is exceeded.
+  """
+  logging.debug('Waiting for {} to open'.format(http_port))
+  operation = operations[operation_id]
```

Comment:
The same here (is KeyError when accessing operations cache considerably possible)